cranelift/wasm/src/code_translator.rs

```diff
@@ -3024,40 +3018,64 @@ fn pop2_with_bitcast(
     (bitcast_a, bitcast_b)
 }
 
-/// A helper for bitcasting a sequence of values (e.g. function arguments). If a value is a
-/// vector type that does not match its expected type, this will modify the value in place to point
-/// to the result of a `raw_bitcast`. This conversion is necessary to translate Wasm code that
-/// uses `V128` as function parameters (or implicitly in block parameters) and still use specific
-/// CLIF types (e.g. `I32X4`) in the function body.
-pub fn bitcast_arguments(
-    arguments: &mut [Value],
-    expected_types: &[Type],
-    builder: &mut FunctionBuilder,
-) {
-    assert_eq!(arguments.len(), expected_types.len());
-    for (i, t) in expected_types.iter().enumerate() {
-        if t.is_vector() {
+fn bitcast_arguments<'a>(
+    builder: &FunctionBuilder,
+    arguments: &'a mut [Value],
+    params: &[ir::AbiParam],
+    is_wasm: impl Fn(usize) -> bool,
+) -> Vec<(Type, &'a mut Value)> {
+    // zip_eq is from the itertools::Itertools trait
+    params
```

Comment:
Can we split up this iterator chain a bit and give names to intermediates to make it easier to follow? I'm thinking perhaps `wasm_param_tys` for the first part up to grabbing `value_type`s; then another intermediate after `zip_eq`, with a comment that we're zipping with given wasm arg values and asserting length is equal; then do the last two filters on those, commenting that we're taking all vectors whose arg-type is not equal to the param's value type.

Reply:
The `SmallVec` in this case will be on-stack so we can be a bit looser with its size I think (and `sub rsp, 128` is almost always faster than `malloc(128)`). For some cases like this I have just picked a reasonable value but to be more objective about it we could try 1, 4, 16 and see what the difference is...